### ISSUE TITLE
Why use this module over encodeURIComponent?

### ISSUE THREAD

**Issue by matthieusieben** [RGBboy/urlsafe-base64#13]
What does this module add as functionality that a simple `encodeURIComponent` doesn't, except and extra dependency ?

**Comment by RGBboy** [comment]
Mate, I changed your issue name to be more constructive. 

As for why you may want to use this module. This module always encodes bytes to base64 encoding in a url safe way. `encodeURIComponent` encodes a string to url safe UTF-8 encoding, which in a lot of cases will end up taking up more bytes. URIs have a maximum length, so if space is important it is better to use this module. This module also enables node.js to encode in the same way.

**Comment by matthieusieben** [comment]
*TL;DR* Useless and confusing

- Encoding base64 to utf8 does not add bytes as base64 string only contains ASCII chars. Even if it did, the result of both `encodeURIComponent` and `urlsafe-base64` will always be used in a larger document (e.g. HTTP query, HTML page) where their encoding will be the same (the one of the document).
- Base64 encoded values should never be used in the **path** part of URLs (this is just common sense if your api respects REST principles), so:
  - base64 encoded values *should* only be used as query string values
  - there is not real need to use `encodeURIComponent` around base64 encoded values when used in query string. So the URL will not be longer than if the value was encoded using this module.
  - Potential `/` contained in the base64 encoded value should not be an issue. They can only appear in the query string where they won't interfere.
  - Even the `=` padding should not be an issue in a query string value.
  - So in the end, the most benefit one can obtain from this modules is potentially 4 bytes of padding. (BTW lots of internet services use raw base64 values in their URL, see the use of JWTs in oAuth for example). If you are 4 bytes short in a URL with a max recommended length of 4kb then you are probably generating too big base64 values anyways.
- The value generated by this module look like base64 but aren't really, making them hard to debug using common base64 decoding tools

**Comment by matthieusieben** [comment]
You don't mention any good practice, that `encodeURIComponent` is probably better fit, and that this modules is probably not even required in most cases, which means that this module contributes to misinform people trying code for the internet.

I actually came across this module when reviewing the code of a junior dev.

**Comment by RGBboy** [comment]
This module solves a problem I had. I decided to open source it. Others seem to be using it (2,649 downloads a day). If this does not meet your needs by all means use `encodeURIComponent` or something else. As for this module, I'll keep using it as it meets the design constraints of the application I built.

To address your comments:

> Encoding base64 to utf8 does not add bytes as base64 string only contains ASCII chars. Even if it did, the result of both encodeURIComponent and urlsafe-base64 will always be used in a larger document (e.g. HTTP query, HTML page) where their encoding will be the same (the one of the document).

I think there may be some confusion here and some assumptions on your end. This module takes a buffer of arbitrary data (any bytes) and encodes them to base64 in a URLSafe way. The `Buffer` part is important as not all sequences of bytes can be represented as utf8 strings. `encodeURIComponent` only accepts strings. So if I want to achieve the same effect using `encodeURIComponent` I would first need to convert the buffer to a string using base64 encoding, then pass the result to `encodeURIComponent`. Node does not encode base64 in a url safe way so the result of passing the string to `encodeURIComponent` results in a larger string than what this module does. For example:

```
Buffer -> URLSafeBase64.encode:
'fBOgDph1i8gWBQrQ1IpnLRZSHMO_2EfTQtFTh4x_lUo' // 43 characters
```

```
Buffer -> toString('base64') -> encodeURIComponent:
'fBOgDph1i8gWBQrQ1IpnLRZSHMO%2F2EfTQtFTh4x%2FlUo%3D' // 50 characters
```

As demonstrated above, `encodeURIComponent` can result in a larger string being produced. If this is a design concern then this module may make sense to use.

> Base64 encoded values should never be used in the path part of URLs (this is just common sense if your api respects REST principles), so:
>  * base64 encoded values should only be used as query string values
>  * there is not real need to use encodeURIComponent around base64 encoded values when used in query string. So the URL will not be longer than if the value was encoded using this module.
>  * Potential / contained in the base64 encoded value should not be an issue. They can only appear in the query string where they won't interfere.
>  * Even the = padding should not be an issue in a query string value.
>  * So in the end, the most benefit one can obtain from this modules is potentially 4 bytes of padding. (BTW lots of internet services use raw base64 values in their URL, see the use of JWTs in oAuth for example). If you are 4 bytes short in a URL with a max recommended length of 4kb then you are probably generating too big base64 values anyways.

Again, I think you have made some assumptions here (which in fairness may apply to the problem `encodeURIComponent` solves for you). Only using base64 values in the query string may be a practice you follow but is not a rule everyone must follow. This module makes no assumption where the output string will be used within a URL nor if the end user is following REST principles. It therefore encodes correctly so the result can be used anywhere within a URL. For my use case, I needed to squeeze as much space out of the URL as possible so this module was created to meet that requirement.

> The value generated by this module look like base64 but aren't really, making them hard to debug using common base64 decoding tools

This is a by product of those tools not supporting base64 encoding to RFC 4648. Unfortunately there is not much I can do about that. Maybe you can raise issues with those tools, or be a nice participant in the open source community and submit PRs to fix ;)

**Comment by RGBboy** [comment]
As for addressing your issue, I have no problem updating the `readme.md` to detail the intention of this module and reference `encodeURIComponent` for users wanting to encode strings in a URL safe way. I am however hesitant of mentioning "good practice" as that is highly subjective.

I think you would do well to reflect on how you approached this issue. I found it offensive, very unconstructive and hard to understand what your underlying issue actually was.

I would have found it much better if you outlined your problem. E.g:

> I found one of my team members using this module when `encodeURIComponent` would have sufficed. We only need to encode the query string of our URLs and therefore can allow `+`, `/` and `=` characters.

And propose a solution. E.g.:

> To avoid this issue I think it would be good to outline the intention of this module in more detail and reference `encodeURIComponent` for those that want to encode query strings.

Be nice and help move things forward.
